### ISSUE TITLE
endianness issue?

### ISSUE THREAD

**Issue by SBFspot** [SBFspot/SBFspot#78]
Hi,

I cross-compiled sbfspot for my MIPS-based openwrt router.  It compiles and runs, but enters an endless loop after establishing a bluetooth connection to the inverter:
Connecting to 00:80:25:2E:0A:5B (1/10)
Initializing...
SUSyID: 125 - SN: 952645759 (0x38C8387F)
getPacket(2)
MAX_CommBuf is now 18 bytes
Received 18 bytes
Received 13 bytes
--------: 00 01 02 03 04 05 06 07 08 09
00000000: 7E 1F 00 61 5B 0A 2E 25 80 00 
00000010: 00 00 00 00 00 00 02 00 00 04 
00000020: 70 00 03 00 00 00 00 01 00 00 
00000030: 00 
cmd=512
Received 18 bytes
Received 13 bytes
--------: 00 01 02 03 04 05 06 07 08 09
00000000: 7E 1F 00 61 5B 0A 2E 25 80 00 
00000010: 00 00 00 00 00 00 02 00 00 04 
00000020: 70 00 03 00 00 00 00 01 00 00 
00000030: 00 
cmd=512
...
(it just keeps retrying)

A connection to the same inverter, using the same USB bluetooth dongle, but from an x86_64 linux pc looks like this:
Connecting to 00:80:25:2E:0A:5B (1/10)
Initializing...
SUSyID: 125 - SN: 963191254 (0x396921D6)
getPacket(2)
MAX_CommBuf is now 18 bytes
Received 18 bytes
Received 13 bytes
--------: 00 01 02 03 04 05 06 07 08 09
00000000: 7E 1F 00 61 5B 0A 2E 25 80 00 
00000010: 00 00 00 00 00 00 02 00 00 04 
00000020: 70 00 03 00 00 00 00 01 00 00 
00000030: 00 
cmd=2
<<<====== Content of pcktBuf =======>>>
--------: 00 01 02 03 04 05 06 07 08 09
00000000: 7E 1F 00 61 5B 0A 2E 25 80 00 
00000010: 00 00 00 00 00 00 02 00 00 04 
00000020: 70 00 03 00 00 00 00 01 00 00 
00000030: 00 
<<<=================================>>>
MAX_pcktBuf is now 31 bytes
...
(successful transfer)

I'm not familiar with the contents of these packets at all, but the cmd=512 (0x0200) vs. cmd=2 (0x0002) makes me think of an endianness problem?  (as the openwrt mips platform is big endian)

#### This work item was migrated from CodePlex
CodePlex work item ID: '73'
Vote count: '1'

**Comment by SBFspot** [comment]
[SBF@2/11/2014]
It's indeed an endianness issue. This has been solved by Claas [some time ago](https://groups.google.com/forum/#!msg/sma-spot/MAwMkGlO99w/9JSH5XFylzAJ) but it never got into the release version

In short, it should work with these changes to SBFnet.cpp:
``` C
//Add
int is_big_endian(void)
{
    union
    {
    uint32_t i;
    char c[4];
    } bint = {0x01020304};

    return bint.c[0] == 1; 
}

//Modify
int64_t get_longlong(BYTE *buf)
{
    register int64_t lnglng = 0;

	if (is_big_endian())
	{
		lnglng += *(buf);
		lnglng <<= 8;
		lnglng += *(buf+1);
		lnglng <<= 8;
		lnglng += *(buf+2);
		lnglng <<= 8;
		lnglng += *(buf+3);
		lnglng <<= 8;
		lnglng += *(buf+4);
		lnglng <<= 8;
		lnglng += *(buf+5);
		lnglng <<= 8;
		lnglng += *(buf+6);
		lnglng <<= 8;
		lnglng += *(buf+7);
	}
	else
	{
		lnglng += *(buf+7);
		lnglng <<= 8;
		lnglng += *(buf+6);
		lnglng <<= 8;
		lnglng += *(buf+5);
		lnglng <<= 8;
		lnglng += *(buf+4);
		lnglng <<= 8;
		lnglng += *(buf+3);
		lnglng <<= 8;
		lnglng += *(buf+2);
		lnglng <<= 8;
		lnglng += *(buf+1);
		lnglng <<= 8;
		lnglng += *(buf);
	}

    return lnglng;
}

int32_t get_long(BYTE *buf)
{
    register int32_t lng = 0;

	if (is_big_endian())
	{
		lng += *(buf);
		lng <<= 8;
		lng += *(buf+1);
		lng <<= 8;
		lng += *(buf+2);
		lng <<= 8;
		lng += *(buf+3);
	}
	else
	{
		lng += *(buf+3);
		lng <<= 8;
		lng += *(buf+2);
		lng <<= 8;
		lng += *(buf+1);
		lng <<= 8;
		lng += *(buf);
	}
	
    return lng;
}

short get_short(BYTE *buf)
{
    register short shrt = 0;

	if (is_big_endian())
	{
		shrt += *(buf);
		shrt <<= 8;
		shrt += *(buf+1);
	}
	else
	{
		shrt += *(buf+1);
		shrt <<= 8;
		shrt += *(buf);
	}

    return shrt;
}
```

Please let us know if it's OK

**Comment by SBFspot** [comment]
[ghen12@2/11/2014]
Hi,
Unfortunately, it's still stuck in the same cmd=512 loop with this patch.

PS: an endianness test could be done at compile (preprocessor) time?  No need to evaluate it at runtime.

**Comment by SBFspot** [comment]
[ghen12@2/11/2014]
I applied (by hand) Claas' original patch, and that works!

But I agree with you there should be a cleaner way.  Could you look into bluetooth.h functions htobs, btohs, and friends?

**Comment by SBFspot** [comment]
[ghen12@3/11/2014]
Just applying btohs macro (bluetooth-to-host for short) seems to work fine:

--- SBFspot/SBFspot.cpp.orig	2014-09-11 21:09:48.000000000 +0200
+++ SBFspot/SBFspot.cpp	2014-11-03 21:33:18.967838798 +0100
@@ -684,9 +684,9 @@
         }
 
         //More data after header?
-        if (pkHdr->pkLength > sizeof(pkHeader))
+        if (btohs(pkHdr->pkLength) > sizeof(pkHeader))
         {
-            bib += bthRead(CommBuf + sizeof(pkHeader), pkHdr->pkLength - sizeof(pkHeader));
+            bib += bthRead(CommBuf + sizeof(pkHeader), btohs(pkHdr->pkLength) - sizeof(pkHeader));
 
             if (DEBUG_HIGH) HexDump(CommBuf, bib, 10);
 
@@ -694,7 +694,7 @@
             if (isValidSender(senderaddr, pkHdr->SourceAddr) == 1)
             {
                 rc = E_OK;
-                if (DEBUG_NORMAL) printf("cmd=%d\n", pkHdr->command);
+                if (DEBUG_NORMAL) printf("cmd=%d\n", btohs(pkHdr->command));
 
                 if ((hasL2pckt == 0) && (CommBuf[18] == 0x7E) && (get_long(CommBuf+19) == 0x656003FF))
                 {
@@ -707,9 +707,9 @@
                     int bufptr = sizeof(pkHeader);
                     bool escNext = false;
 
-                    if (DEBUG_NORMAL) printf("PacketLength=%d\n", pkHdr->pkLength);
+                    if (DEBUG_NORMAL) printf("PacketLength=%d\n", btohs(pkHdr->pkLength));
 
-                    for (int i = sizeof(pkHeader); i < pkHdr->pkLength; i++)
+                    for (int i = sizeof(pkHeader); i < btohs(pkHdr->pkLength); i++)
                     {
                         pcktBuf[index] = CommBuf[bufptr++];
                         //Keep 1st byte raw unescaped 0x7E
@@ -760,7 +760,7 @@
             if (isValidSender(senderaddr, pkHdr->SourceAddr) == 1)
             {
                 rc = E_OK;
-                if (DEBUG_NORMAL) printf("cmd=%d\n", pkHdr->command);
+                if (DEBUG_NORMAL) printf("cmd=%d\n", btohs(pkHdr->command));
 
                 memcpy(pcktBuf, CommBuf, bib);
                 packetposition = bib;
@@ -780,7 +780,7 @@
         }
     }
     // changed to have "any" wait4Command (0xFF) - if you have different order of commands
-    while (((pkHdr->command != wait4Command) || (rc == E_RETRY)) && (0xFF != wait4Command));
+    while (((btohs(pkHdr->command) != wait4Command) || (rc == E_RETRY)) && (0xFF != wait4Command));
 
     if ((rc == E_OK) && (DEBUG_HIGH))
     {

**Comment by SBFspot** [comment]
[ghen12@3/11/2014]
Just applying btohs macro (bluetooth-to-host for short) seems to work fine:
```
--- SBFspot/SBFspot.cpp.orig	2014-09-11 21:09:48.000000000 +0200
+++ SBFspot/SBFspot.cpp	2014-11-03 21:33:18.967838798 +0100
@@ -684,9 +684,9 @@
         }
 
         //More data after header?
-        if (pkHdr->pkLength > sizeof(pkHeader))
+        if (btohs(pkHdr->pkLength) > sizeof(pkHeader))
         {
-            bib += bthRead(CommBuf + sizeof(pkHeader), pkHdr->pkLength - sizeof(pkHeader));
+            bib += bthRead(CommBuf + sizeof(pkHeader), btohs(pkHdr->pkLength) - sizeof(pkHeader));
 
             if (DEBUG_HIGH) HexDump(CommBuf, bib, 10);
 
@@ -694,7 +694,7 @@
             if (isValidSender(senderaddr, pkHdr->SourceAddr) == 1)
             {
                 rc = E_OK;
-                if (DEBUG_NORMAL) printf("cmd=%d\n", pkHdr->command);
+                if (DEBUG_NORMAL) printf("cmd=%d\n", btohs(pkHdr->command));
 
                 if ((hasL2pckt == 0) && (CommBuf[18] == 0x7E) && (get_long(CommBuf+19) == 0x656003FF))
                 {
@@ -707,9 +707,9 @@
                     int bufptr = sizeof(pkHeader);
                     bool escNext = false;
 
-                    if (DEBUG_NORMAL) printf("PacketLength=%d\n", pkHdr->pkLength);
+                    if (DEBUG_NORMAL) printf("PacketLength=%d\n", btohs(pkHdr->pkLength));
 
-                    for (int i = sizeof(pkHeader); i < pkHdr->pkLength; i++)
+                    for (int i = sizeof(pkHeader); i < btohs(pkHdr->pkLength); i++)
                     {
                         pcktBuf[index] = CommBuf[bufptr++];
                         //Keep 1st byte raw unescaped 0x7E
@@ -760,7 +760,7 @@
             if (isValidSender(senderaddr, pkHdr->SourceAddr) == 1)
             {
                 rc = E_OK;
-                if (DEBUG_NORMAL) printf("cmd=%d\n", pkHdr->command);
+                if (DEBUG_NORMAL) printf("cmd=%d\n", btohs(pkHdr->command));
 
                 memcpy(pcktBuf, CommBuf, bib);
                 packetposition = bib;
@@ -780,7 +780,7 @@
         }
     }
     // changed to have "any" wait4Command (0xFF) - if you have different order of commands
-    while (((pkHdr->command != wait4Command) || (rc == E_RETRY)) && (0xFF != wait4Command));
+    while (((btohs(pkHdr->command) != wait4Command) || (rc == E_RETRY)) && (0xFF != wait4Command));
 
     if ((rc == E_OK) && (DEBUG_HIGH))
     {
```

**Comment by SBFspot** [comment]
[SBF@4/11/2014]
I was a little bit confused... only the structs are affected.
Need some changes for windows too (btohs doesn't exist)

**Comment by SBFspot** [comment]
**Issue closed by _SBF_ with comment**
Fixed in 3.0.2

**Reason closed**
Fixed
